teps/0002-custom-tasks.md

```diff
@@ -0,0 +1,392 @@
+
+This Pipeline pipes its input param `"pl-wait-duration"` to the `example-task`:
+
+```
+apiVersion: tekton.dev/v1beta1
+kind: Pipeline
+metadata:
+  name: example
+spec:
+  params:
+  - name: pipeline-param
+    default: hello
+  tasks:
+  - name: example-task
+    taskRef:
+      apiVersion: example.dev/v0
+      kind: Example
+      name: my-example
+    params:
+    - name: task-param
+      value: "$(params.pipeline-param)"
+```
+
+In this case, we would expect the `Example` CRD type accepts a parameter named `task-params`, expressed by having a field `.spec.params`, conformant to the `ParamSpec` type.
+
+```
+apiVersion: example.dev/v0
+kind: Example
+metadata:
+  name: my-example
+spec:
+  params:
+  - name: task-param
+    default: goodbye
+  someString: "$(params.task-param), old friend."
+```
+
+When run, the `PipelineRun` controller will create a `Run<Example>`, passing in either the `Pipeline`'s parameter default (`hello`) or the `PipelineRun`-specified override value, if present:
+
+```
+apiVersion: tekton.dev/v1alpha1
+kind: Run
+metadata:
+  name: run-<tekton-generated-suffix>
+spec:
+  ref:
+    apiVersion: example.dev/v0
+    kind: Example
+    name: my-example
+  params:
+  - name: task-param
+    value: hello
+```
+
+The Custom Task author's controller will be watching for creations of `Run`s of this type, and can invoke a Tekton-provided Go package to resolve parameters in string-typed fields, using reflection. This will give them a struct value like:
+
+```
+{
+  "spec": {
+    "params": [ … ],
+    "task-param": "hello, old friend"
+  }
+}
+```
+
+### Result Reporting
+
+Custom Task controllers can report any results via the `Run`'s `.status.results` field:
+
+```
+apiVersion: tekton.dev/v1alpha1
+kind: Run
+metadata:
+  name: run-blah-blah
+spec:
+  … 
+status:
+  results:
+  - name: first-name
+    value: Bob
+```
+
+If their custom defined CRD type has a `.results` field of a compatible type, `Pipeline` validation can take that into account when validating inputs and outputs between Tasks (custom or traditional) in a `Pipeline`.
+
+Controllers can report any results, regardless of whether the underlying CRD object declared them.
+
+### Cancellation
+
+To support cancellation of `Run`s, when a `PipelineRun` is cancelled, the `PipelineRun`  controller will attempt to update any ongoing `Run`s' `.spec.status` to `"Cancelled"`, and update `.status.conditions` to signal unsuccesful execution.
+
+A Custom Task author can watch for this status update and take any corresponding actions (e.g., cancel a cloud build, stop the waiting timer, tear down the approval listener).
+
+Supporting cancellation is optional but recommended.
+
+### Timeout
```

Comment:
I feel like another thing we need to plumb through in general is identity e.g. `serviceAccountName`

Reply:
That's one we'd need to plumb through for any Custom Task that intends to end up running Pod-based workflows (e.g., MatrixTask, GroupTask), but not the general case.

In the meantime a MatrixTask config can specify and use `serviceAccountName` and it just won't be overrideable from a Pipeline. Perhaps it should just be a param? (e.g., `serviceAccountName: $(params.serviceAccountName)`)

Reply:
I think this is a usability issue.  Users now have to maintain the service account name setting in multiple places (the existing PR serviceAccountName keys and now separate task parameters which could be custom-task specific).  People will question why the pipelinerun's configured service account name can't just be passed down.

Reply:
Should we have a specific header `serviceAccount` and related subject then ? :upside_down_face: